staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,250 @@
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"math"
+	"sort"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// unbounded uses nil to represent an unbounded cardinality value.
+var unbounded *uint64 = nil
+
+// CelSchemaContext keeps track of data used by x-kubernetes-validations rules for a specific schema node.
+type CelSchemaContext struct {
+	// Schema is a structural schema for this CelSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules or if the schema is not a valid structural schema.
+	Schema *structuralschema.Structural
+	// DeclType is a CEL declaration representation of Schema of this CelSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules or if the schema is not a valid structural schema.
+	DeclType *model.DeclType
+
+	// Converter converts a JSON schema into a structuralschema.Structural and model.DeclType.
+	Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+	// MaxCardinality represents a limit to the number of data elements that can exist for the current
+	// schema based on MaxProperties or MaxItems limits present on parent schemas, If all parent
+	// map and array schemas have MaxProperties or MaxItems limits declared MaxCardinality is
+	// an int pointer representing the product of these limits.  If least one parent map or list schema
+	// does not have a MaxProperties or MaxItems limits set, the MaxCardinality is nil, indicating
+	// that the parent schemas offer no bound to the number of times a data element for the current
+	// schema can exist.
+	MaxCardinality *uint64
+
+	// TotalCost accumulates the x-kubernetes-validators estimated rule cost total for an entire custom resource
+	// definition. A single TotalCost is allocated for each validation call and passed through the stack as the
+	// custom resource definition's OpenAPIv3 schema is recursively validated.
+	TotalCost *TotalCost
+}
+
+// Converter converts from JSON schema to a structural schema and a CEL declType, or returns an error if the conversion
+// fails.
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+// DefaultConverter converts using standard conversion utilities.
+var DefaultConverter = func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error) {
+	structural, err := structuralschema.NewStructural(schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	return structural, model.SchemaDeclType(structural, isRoot), nil
+}
+
+// RootCelContext constructs CelSchemaContext for the given root schema.
```

Comment:
nit: consider using `...CEL...` instead of `...Cel...` everywhere?

Reply:
SGTM. Done.